### DISCUSSION TITLE
Can i make this project using Arduino mega2560 and LidarA1M8 instead of Teensy and A2M8?

### DISCUSSION THREAD

**Discussion by thanhtruongphan** [fjp/2wd-robot#94] [Q&A]
Hi everyone, i'am a newbie ROS, i want to make a ROS mobile robot myself, this project is very helpful but I only have RPi 4B, Arduino Mega2560, Lidar A1M8 now.

Can i do this projet with these? 
if yes, 
how can I do? 

Thank you.

**Comment by fjp** [comment]
Hi @thanhtruongphan thanks for the question.

Yes you can also use A1M8, there is also a mount available for REMO robot:

![image](https://github.com/user-attachments/assets/891561b9-d810-441e-b5c7-119d6e4bbf1c)


| <img src="https://github.com/user-attachments/assets/e7be8f04-788c-456a-892d-13983baf9a6d" width="80%"> | <img src="https://github.com/user-attachments/assets/f2d5e179-f27a-432c-ab2d-c283f15def65" width="80%"> | <img src="https://github.com/user-attachments/assets/3249c3a1-010b-41fc-ba29-348088b3ef8c" width="80%">  | <img src="https://github.com/user-attachments/assets/2e9842e9-ee6e-42eb-8d17-ea9a2095fb90" width="80%"> |
|:---:|:----:|:----:|:---:|


You can get it from [here](https://fjp.gumroad.com/l/GnMpU) or by becoming a sponsor you will receive all the latest updates.

For the lidar ROS node check out: https://github.com/Slamtec/rplidar_ros it also supports A1 lidars.

And instead of flashing the base_controller software onto the teensy you can go for Arudino. The documentation explains the process for teensy but it should be quite similar for Arduino https://ros-mobile-robots.com/processing_units/teensy-mcu/#base-controller

**Reply by thanhtruongphan** [reply]
yeah! 

thanks for your reply. I will see the documentation explains that you've attached.

beside, I see the Part List, it has a Groove I2C driver and a Boosch IMU, but I certainly don't have these. 

all i have are
- raspi4 4GB
- arduino mega 2560
- pi camera v2
- IMU name's MPU6050
- Lidar A1M8
- 2 motor with encoder
- power 12v
- power 5v for rasp, 
- L298 motor driver 

I had build those to a diff-drive robot base, hardware is simple and no problem. 
Can I apply your software-source-code to my hardware? If not, I want to change something to cooperate with my ideas, but I don't know how to do...

and, I really want to know each part of your source, but there are very much documentations that make me confused.

can you give me some instruction to learn?

Thanks a lot...!!!

**Reply by fjp** [reply]
## Motor Driver

The Groove I2c driver was used for the first version of diffbot software were most of the motor control was done on a high level running on the single board computer and only the code to count the encoder ticks was running on the microcontroller (see this [deprecated encoders folder](https://github.com/ros-mobile-robots/diffbot/tree/noetic-devel/diffbot_base/scripts/encoders)). With the latest state of diffbot I suggest to run the motor driver software on the microcontroller board (Arduino in your case) and have a look into the `base_controller` low level scripts in the [diffbot_base package](https://github.com/ros-mobile-robots/diffbot/tree/noetic-devel/diffbot_base/scripts/base_controller/lib). 

![image](https://github.com/user-attachments/assets/df4b5ff5-7574-47e8-b421-d9af310baf13)

REMO robot, the latest version, uses [Adafruit feather wing motor driver](https://github.com/ros-mobile-robots/diffbot/tree/noetic-devel/diffbot_base/scripts/base_controller/lib/motor_controllers/adafruit_feather_wing) which implements the simple [motor_controller_interface](https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_base/scripts/base_controller/lib/motor_controller_interface/motor_controller_interface.h). In case you want to use the L298 motor driver I suggest you create a new cpp and header file similar to the one as [Adafruit feather wing motor driver](https://github.com/ros-mobile-robots/diffbot/tree/noetic-devel/diffbot_base/scripts/base_controller/lib/motor_controllers/adafruit_feather_wing). Once that is done you can use it in the [main.cpp](https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_base/scripts/base_controller/src/main.cpp#L12-L13) of the `base_controller` node.

## IMU

In the beginning you don't need the IMU. Everything will work without IMU and run fine with the encoders. When everything works with the encoders as localization input you can add the IMU which does improve the localization, because the encoders can be inaccurate. For the IMU support you would need to look into https://github.com/ros-mobile-robots/diffbot/pull/60 and https://github.com/ros-mobile-robots/diffbot/pull/37. Still needs some testing to be integrated into the source code as a new ros package.

**Comment by thanhtruongphan** [comment]
hiii Mr. Franz, 

I just launch with "roslaunch diffbot_bringup bringup_with_laser.launch"

everything is okay, but It has a problem, (as the picture i attached)

I have tried to change baudrate to 57600, 9600 but it still error.

help me :D 

thank u so much!

![Screenshot 2024-10-02 211649](https://github.com/user-attachments/assets/20f65c49-d8ac-4330-bb4b-6f17c0613c56)

**Reply by fjp** [reply]
Hi @thanhtruongphan, have you gone through these steps: https://ros-mobile-robots.com/diffbot_msgs/#using-rosmsg? Also please check the references mentioned on this page such as https://wiki.ros.org/rosserial_arduino/Tutorials/Arduino%20IDE%20Setup#Install_ros_lib_into_the_Arduino_Environment.

Have you tried with a simpler example first to see if that works? Might be something in the code that isn't working as expected on the Arduino (compared to Teensy).

**Reply by thanhtruongphan** [reply]
yeah, 

of course i had done follow 
> mentioned on this page such as https://wiki.ros.org/rosserial_arduino/Tutorials/Arduino%20IDE%20Setup#Install_ros_lib_into_the_Arduino_Environment.

and I have tried 
>>> ![Screenshot 2024-10-02 231758](https://github.com/user-attachments/assets/cd6b14f0-6f8f-4e11-bdd3-dce7dbfff61e)
look like it's ok...

> 
> Have you tried with a simpler example first to see if that works? Might be something in the code that isn't working as expected on the Arduino (compared to Teensy).
 - yeah, but Teensy isn't available where i live... :']

**Reply by fjp** [reply]
Just to confirm: is it working with one of the rosserial lib provided examples? Have you run https://wiki.ros.org/rosserial_arduino/Tutorials/Hello%20World and https://wiki.ros.org/rosserial_arduino/Tutorials/Blink both successfully or do you see the same error?

Are you using Arduino IDE or PlatformIO?

Here is a config that should get a successful build with PlatformIO, although I had to change `std::memcpy` to `memcpy`:

```
[env:megaatmega2560]
platform = atmelavr
board = megaatmega2560
framework = arduino
lib_deps = 
    frankjoshua/Rosserial Arduino Library@^0.9.1
    ./lib
    adafruit/Adafruit Motor Shield V2 Library@^1.0.11
    Wire
    Encoder
```

You could also try to run a simpler code example similar to the following unit test to manually check the encoders are working and also rosserial together with the Arduino doesn't produce any errors when connected to the Raspberry Pi SBC: https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_base%2Fscripts%2Fbase_controller%2Ftest%2Fencoders%2Ftest_encoders.cpp

Regarding the baudrate, where did you change it and have you also tried to set it higher than 57600? These instructions might be useful: https://wiki.ros.org/rosserial_arduino/Tutorials/NodeHandle%20and%20ArduinoHardware

If that all doesn't work I can try to reproduce it on an Arduino Mega.

Can you also please share your motor_l298_controller implementation so we can check it? Can also be through a PR.

Finally, sorry for all the questions 😅 I am just trying to get as much info so we find a solution.

**Reply by fjp** [reply]
You probably have found this issue already: https://answers.ros.org/question/210875/arduino-rosserial-unable-to-sync-with-device/? I just want to highlight some steps that might be worth checking:

- is the `rosrun ... _port:=?` correct?
- add a `Serial.begin()` and `nh.getHardware()->setBaud()` (same baud as used for the `rosrun ... _baud:=` command.

![image](https://github.com/user-attachments/assets/26917bba-32c5-4de6-ba29-aa0b697abe8c)

![image](https://github.com/user-attachments/assets/f38a1e8e-951a-4a4e-8c87-ef567785ea77)

![image](https://github.com/user-attachments/assets/f1652a58-4f0c-4280-ba78-dc20d2066544)

**Reply by thanhtruongphan** [reply]
yeah, I have tried everything you suggested above, 

finally, this error is fixed. 
but there are some new problem occur, haha 

like this image, i tried 115200, 57600, 9600, all is same... 
>>> ![image](https://github.com/user-attachments/assets/8d253fe2-c11c-4433-ae8b-b70608ceca5e)



> If that all doesn't work I can try to reproduce it on an Arduino Mega.
> 
> Can you also please share your motor_l298_controller implementation so we can check it? Can also be through a PR.

of course, and I have a backup on github, for convenience, here is link: (choose branch Master :D) (... I accidentally overwrote the new folder without knowing, the edited content was deleted, I will update it soon)... https://github.com/thanhtruongphan/diffbot_01.git

all i modified in here: 
>>> >>![image](https://github.com/user-attachments/assets/d49b811f-0527-402f-9db6-a79f7ba3c7b5)

> 
> Finally, sorry for all the questions 😅 I am just trying to get as much info so we find a solution.

all your questions are really helpful for me... i really appreciate it.
Hope you can help me to solve all soon :D 

thank you so much.

**Reply by fjp** [reply]
Thanks for sharing your code 👍 

> yeah, I have tried everything you suggested above,
> finally, this error is fixed.

Good to hear that you could fix one error. Do you know which changes in particular did help to get over the `Unable to sync with device` error? Would be great if you could share it, so that it helps others that encounter the same error.

The issues you mention sound like memory issue. Comparing [Arduino Mega](https://docs.platformio.org/en/latest/boards/atmelavr/megaatmega2560.html?utm_source=platformio&utm_medium=piohome) and [Teensy](https://docs.platformio.org/en/latest/boards/teensy/teensy40.html):

| Microcontroller | ATMEGA2560 | | IMXRT1062 |
| -- | -- | -- | --|
| Frequency | 16MHz | | 600MHz |
| Flash | 248KB | |  1.94MB |
| RAM | 8KB | |  512KB |
| Vendor | Arduino | |  Teensy |

It is possible that the firmware is too big for Arduino Mega, which is also mentioned at https://github.com/ros-drivers/rosserial/issues/125#issuecomment-355403278:

> In my experience, any Arduino with less than 40KB of memory (e.g. Uno, Leo, etc) will not work with Rosserial, as the relatively large packet sizes overwhelm the Arduino's memory, and cause the "Unable to sync with device".

According to https://github.com/yoneken/rosserial_stm32/issues/47:
> This message often appears when messages are sent frequently, so could you try adding a wait after sending messages?
The problem may also be resolved by increasing the buffer size.

You could try decreasing the frequency with which messages are sent and remove as much unneeded code as possible (debugging messages for example just to see if it helps). If all that doesn't help then I don't think you will be able to use Arudino Mega together with rosserial 😢  Sorry I didn't check the Arduino Mega RAM earlier, I wasn't aware that this could be an issue.

But I want to suggest one other approach that should work according to [andino](https://github.com/Ekumen-OS/andino), a ROS2 based robot which is quite similar to diffbot project and even took some inspiration from it. They don't use rosserial (or micro ros which is the ROS2 equivalent more or less). Instead they use https://libserial.readthedocs.io/en/latest/description.html running on the SBC and interacting with the microcontroller through USB and a quite simple but efficient command interface, inspired from https://github.com/hbrobotics/ros_arduino_bridge/blob/indigo-devel/ros_arduino_firmware/src/libraries/ROSArduinoBridge/ROSArduinoBridge.ino

Here is the higher level C++ code running on the SBC: https://github.com/Ekumen-OS/andino/blob/humble/andino_base/src/motor_driver.cpp that interacts with the lower level code running on the arduino microcontroller. As interface they use simple serial commands, see [commands.h](https://github.com/Ekumen-OS/andino/blob/humble/andino_firmware/src/commands.h). Here are the most used commands:

- Get encoder values: 'e'
- Set open-loop speed for the motors[pwm] 'o <left> <right>'
  Example to move forward full speed: 'o 255 255'
  Range [-255 -> 255]
- Set closed-loop speed for the motors[ticks/sec] 'm <left> <right>'
- Set PID values: 'u <kp> <kd> <ki> <offset>'

If the suggestions above don't bring any solution and you don't plan (or can as you mention) to upgrade to another microcontroller, I only see that you implement a similar approach as andino (ROS2). This obviously requires a bit more work if you want to still use the rest of diffbot (ROS1):

1. Take over (adapt) [`andino_firmware`](https://github.com/Ekumen-OS/andino/tree/humble/andino_firmware) code to your needs
2. Adapt `diffbot_base` ros control hw interface (write method and receiving encoder ticks) to work with the serial interface from andino
    1. Instead of publishing the motor commands via ros you would interact with the microcontroller through libserial, see this part of the [diffbot_hw_interface.cpp](https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_base/src/diffbot_hw_interface.cpp#L226-L227)
    2. Read encoder ticks in the [read method](https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_base/src/diffbot_hw_interface.cpp#L136-L161) via libserial (andino's `e` command, called [here in their motor_driver.cpp](https://github.com/Ekumen-OS/andino/blob/humble/andino_base/src/motor_driver.cpp#L65-L73)). Then you don't need the [callback here](https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_base/src/diffbot_hw_interface.cpp#L358-L365) that obtains the ticks via ros

**Reply by thanhtruongphan** [reply]
continue modify Arduino with ROS is so hard... yeah, may be i will get a lot of problem. I will think about that. 

oh i just have noticed that STM32 is quite same with teensy about hardware, and it's available in my region. 
Do you think it's easy for me to continue do it with this project? 

pls tell me your ideas and give me some instructions if you have experiences with it. 

i just think about that :D 

many thanks and have a nice day!

**Reply by fjp** [reply]
> oh i just have noticed that STM32 is quite same with teensy about hardware, and it's available in my region.
Do you think it's easy for me to continue do it with this project?

> pls tell me your ideas and give me some instructions if you have experiences with it.

If you can find and get a STM32 board that has similar technical specs and I/Os as the Teensy 4.0 I think it is worth a try. The list of STM32 boards seems quite long: https://docs.platformio.org/en/latest/boards/index.html#st-stm32 and it looks like Teensy and STM32 are similar.

**Reply by thanhtruongphan** [reply]
hi @fjp , I have tried with STM32F401RCT6 (64kb SRAM), i modified all files in 'Base_controller' folder but still the same functions like the original.
Now i can communicate with rosserial without error mismatch, and when I try the command:
> ```roslaunch diffbot_bringup bringup.launch```

Result: (maybe it doesn't find /cmd_vel to subscribe, when no Command velocity input, it 'Emmercency Stop')
>>>>> ![image](https://github.com/user-attachments/assets/82626fe6-a60b-4963-8650-28b7630555df)

And, I've just noticed that I can't find the "**subscribe to /cmd_vel**" anywhere in your source code: 
>>> ![image](https://github.com/user-attachments/assets/578f6f60-71b0-4ab3-a15a-f720e9bd9663)
when the Docs said: 
>>> ![image](https://github.com/user-attachments/assets/9e486ef8-37f0-453a-8ea6-f5b36bcdb64d)

hmmm, I wanted to share what I tried, and what did you think, I wanted to ask for your feedback. 
help me :D 


thank you so much!!!

**Reply by thanhtruongphan** [reply]
>> ![Screenshot 2024-10-21 171236](https://github.com/user-attachments/assets/2a93189c-6514-41ac-b261-4f1bbe897507)
hope to your reply soon... Because I'm not know how to continue now... 👯

**Reply by fjp** [reply]
Hi @thanhtruongphan great work in getting everything running on the new mcu 👍 

The [`diffbot_bringup keyboard_teleop.launch`](https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_bringup/launch/keyboard_teleop.launch#L4) is just used for debugging purposes and will only perform a `<remap from="/cmd_vel" to="/diffbot/mobile_base_controller/cmd_vel"/>` but doesn't actually start crucial robot code. First I recommend you run [`roslaunch diffbot_bringup bringup_with_laser.launch`](https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_bringup/launch/bringup_with_laser.launch) on the SBC of the robot. This will use the [`minimal.launch`](https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_bringup/launch/minimal.launch) that does most of the important bring up. Then you can check for the cmd_vel topic using `rostopic list`. When you see it you should also be able to use the teleop launch file to control your robot using the keyboard. And also the slam and navigation stack running from your development machine using the command: 

First, brinup the robot hardware including its laser with the following launch file from the [diffbot_bringup](https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_bringup) package. Make sure to run this on the real robot (e.g. connect to it via ssh):

```
roslaunch diffbot_bringup bringup_with_laser.launch
```

Then, in a new terminal on your remote/work development machine (not the single board computer) run the slam gmapping command:

```
roslaunch diffbot_slam diffbot_slam.launch slam_method:=gmapping
```

This should open up RViz and the [rqt_robot_steering](http://wiki.ros.org/rqt_robot_steering) plugin.

Next, steer the robot around manually either using the keyboard_teleop node or using the rqt_robot_steering node and save the map with the following command when you are done exploring:

```
rosrun map_server map_saver -f office
```

After the mapping process it is possible to use the created map for navigation, after running the following launch files:

On the single board computer (e.g. Raspberry Pi) make sure that the following is launched:

```
roslaunch diffbot_bringup bringup_with_laser.launch
```

Then on the work/remote development machine run the diffbot_hw.lauch from the diffbot_navigation package:

```
roslaunch diffbot_navigation diffbot_hw.lauch
```

Among other essential navigation and map server nodes, this will also launch an instance of RViz on your work pc where you can use its tools to:

- Localize the robot with the "2D Pose Estimate" tool (green arrow) in RViz
- Use the "2D Nav Goal" tool in RViz (red arrow) to send goals to the robot

A video that does all this can be found here: https://youtu.be/IcYkQyzUqik?si=Y0l7x5qMnfSJZQhi

**Reply by fjp** [reply]
One more note about the `wrong checksum for topic id and msg` message you see. I experienced this when the delay wasn't high enough, on some code I used in an earlier version: https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_base/scripts/encoders/encoders/encoders.ino#L75-L87

Maybe you need to increase the update rates a bit and see if this helps to get rid of the message: https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_base/scripts/base_controller/lib/config/diffbot_base_config.h#L25-L29

**Reply by thanhtruongphan** [reply]
yeah!

I run command on SBC:
```roslaunch diffbot_bringup bringup_with_laser.launch```
and on dev PC:
```roslaunch diffbot_slam diffbot_slam.launch slam_method:=gmapping``` 

result is: 
![Screenshot 2024-10-22 095021](https://github.com/user-attachments/assets/b784f20e-c43b-4733-b9ac-40d565f71bc3)

But no **rqt_steering** appears.
ps: The "wrong checksum ..." appear sometime, may be because the Board i use (just 64KB SRAM, and 16MHz onboard clock - not 84MHz like the Hardware info in [here.](https://docs.platformio.org/en/latest/boards/ststm32/genericSTM32F401RC.html)


Then i tried ```rosrun rqt_robot_steering rqt_robot_steering``` 
Result: 
![Screenshot 2024-10-22 095315](https://github.com/user-attachments/assets/94048955-ef00-4774-a8ed-c1b9280047cc)

but my motors still does not move :D 

I have tried ```rostopic list```: 
>>>>> ![Screenshot 2024-10-22 095957](https://github.com/user-attachments/assets/92b4ac5d-31d1-48e8-9fc2-975e3f4e5e64)

the /cmd_vel is shown but i don't know where it is to break the code to examine, I think I must break the code to examine part-to-part with my code for stm32.

have a nice day, Fjp 💯

**Reply by thanhtruongphan** [reply]
Ahh... Looks like we have a time zone difference? It's 12:30 pm where I am now. 😄

**Reply by fjp** [reply]
I think you are pretty close to getting it running but the errors you are seeing from rosserial, e.g. "[ERROR]: Mismatched protocol version in packet (b'\x00'): lost sync or rosserial_python is from different ros release than the rosserial client" looks a lot like there is too much running on the mcu. 

I'd try to reduce it as much as possible. For example remove the eStop part completely and only use the encoders part or motor part to see if you can obtain encoder ticks (`rostopic echo`) or send something to the mcu via rosserial when the encoder code is commented out. 

And try to increase the update rates maybe even more https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_base/scripts/base_controller/lib/config/diffbot_base_config.h#L25-L29

Good luck and have a nice day too 😄

**Reply by thanhtruongphan** [reply]
>  For example remove the eStop part completely and only use the encoders part or motor part to see if you can obtain encoder ticks

yeah, of course I'll try. At least I could send something via rosserial now, as you said...
>or send something to the mcu via rosserial when the encoder code is commented out.

but I don't know how to do... I think I'll use teleop-twist-keyboard to send via **/cmd_vel** but I can't find it in the Diffbot-repositories. 

I could modify code for STM32 to same as Teensy, but I don't know how to check it deeply.

Thank Franz 🥹

**Reply by thanhtruongphan** [reply]
![image](https://github.com/user-attachments/assets/4b5e3f54-a976-40ae-b09d-4237c27fbf5c)

as you can see, i try to change Velocities but my motor still not move

**Reply by fjp** [reply]
To me it seems the issue is on the mcu board. But to verify let's try to get some more log outputs and understand where in the control flow you get stuck (cmd_vel -> code running on RPi -> high level ros control hardware interface ([write method](https://github.com/ros-mobile-robots/diffbot/blob/8d4e432c43a8b6f90c99a145ed1c130f83f5de26/diffbot_base/src/diffbot_hw_interface.cpp#L163)) -> publish motor topics -> mcu board -> low level rosserial base_controller code -> subscriber of the wheel command -> translate to motor commands -> motor driver interaction.

You are interested in this [`pub_wheel_cmd_velocities_`](https://github.com/ros-mobile-robots/diffbot/blob/8d4e432c43a8b6f90c99a145ed1c130f83f5de26/diffbot_base/src/diffbot_hw_interface.cpp#L186), which should publish messages to `wheel_cmd_velocities` topic. I suggest you also use `rostopic echo` to see if you get any output on this topic when you try to command the robot around (using rqt_robot_steering or keyboard_teleop). If that works the problem has to be on the mcu board. 

To investigate the mcu board I would add more logging on to the base_controller code that is running on the mcu. For example add some more logging similar to this https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_base/scripts/base_controller/lib/base_controller/base_controller.h#L594. You can try adding some more log messages to the [subscriber code of `wheel_cmd_velocities `](https://github.com/ros-mobile-robots/diffbot/blob/8d4e432c43a8b6f90c99a145ed1c130f83f5de26/diffbot_base/scripts/base_controller/lib/base_controller/base_controller.h#L465-L476) (for example a simple output like "received wheel_cmd_velocities message"). When you see this trying to move around then there is either a logic problem in how you interact with your motor driver or some update rate issues.


Just a side note: you should see more debug messages from the ros control hardware interface (running on the SBC) when you enable logging here: https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_base/config/base.yaml#L11. But it shouldn't be required, you can also add more relevant logging to the code in the write() method without enabling this flag

**Reply by thanhtruongphan** [reply]
Hey Franz, have a nice day! 😄

I'm still not fix it correctly. 

Here's the code base_controller I've modified: [Github](https://github.com/thanhtruongphan/diffbot_0/tree/master)

If you have free time, pls check it with me. Thank you so much!!

And here is the problem now:
- I run command ```roslaunch diffbot_bringup bringup.launch``` and result is all like: 
>>>>>![Screenshot 2024-10-26 095046](https://github.com/user-attachments/assets/6c5d635e-72e1-489f-a6aa-38634cc50b1c)
>>>>>![Screenshot 2024-10-26 095100](https://github.com/user-attachments/assets/e88ae23d-6eeb-410f-a3bf-3ba50b42695b)
>>>>>![Screenshot 2024-10-26 095110](https://github.com/user-attachments/assets/1f11a26c-14f4-4374-9f4b-14b5041f53e1)
>>>>>![Screenshot 2024-10-26 095123](https://github.com/user-attachments/assets/d68fc3e9-73f7-4ae2-b99c-6e810a86a112)
>>>>>![Screenshot 2024-10-26 095135](https://github.com/user-attachments/assets/3889f297-e0e6-46e3-9a59-aba301e981fc)
>>>>>![Screenshot 2024-10-26 095147](https://github.com/user-attachments/assets/9f6ac40f-e002-4b00-91eb-6d9501855462)
>>>>>![Screenshot 2024-10-26 095210](https://github.com/user-attachments/assets/9ce9095d-80b6-47b1-a303-a194157e7f07)
>>>>>![Screenshot 2024-10-26 095229](https://github.com/user-attachments/assets/90b0f111-8362-4df9-8349-74e7694e87ab)

the result **[INFO] [1729910983.080179]: encoder_ticks_:01980** is because i code the [log_info](https://github.com/thanhtruongphan/diffbot_0/blob/1a7bee00b3b092774ea7496c92e900c0bbf94493/src/encoder/encoder_diffbot_mod_stm32.cpp#L58C1-L71C34) and in mainpp file with [getRPM()](https://github.com/thanhtruongphan/diffbot_0/blob/aed8b75960df188a37622851dac70bd2ce301714/src/mainpp.cpp#L95-L96)

And when I turn round the real-wheel, it **[ERROR] [1729910996.407401]: Lost sync with device, restarting...** and cannot continue communicate with ROS.

--
about control PWM, if I uncomment the [two lines](https://github.com/thanhtruongphan/diffbot_0/blob/aed8b75960df188a37622851dac70bd2ce301714/src/mainpp.cpp#L58-L59) on the ```mainpp.cpp``` code, 
it's OK, the left wheel turn with 750/1000, and the right wheel turn with 250/1000 correctly (the max PWM that I configed from hardware is 1kHz). I had test both encoders and PWM on MCU and it's ok. Maybe the base_controller code that I modified have problem...

The MCU that I use is STM32F411CE with 128kb RAM, 512kb Flash and 100MHz [link](https://docs.platformio.org/en/latest/boards/ststm32/genericSTM32F411CE.html)

It's quite large than Teensy 3.2 (of course It's still small than Teensy 4.0 very much), 

Could you have and ideas for me...? 
Thank you 🥹

**Reply by fjp** [reply]
Hi @thanhtruongphan you might have already found the following suggestions but let me share them here as well:

> This error can mean so many different things. First of all you should check that the topic names and message type actually line up. Then check the baud rate. Also this might happen if any of the subscribers or publishers stop existing during the connection. So make sure you never go out of scope. As in you can't have a subscriber stop existing after you send the message.

source: https://www.reddit.com/r/ROS/comments/18m4k36/error_lost_sync_with_device_restarting/

- Maybe you are missing the custom message generation step? https://ros-mobile-robots.com/diffbot_msgs/#rosserial
- Someone mentioned they had to call `nh.negotiateTopics();` to fix the issue `Tried to publish before configured topic id ...`

source: https://answers.ros.org/question/232355/tried-to-publish-before-configured-topic-id-125/

## Code review

I've had a quick look at your code and I am not sure is if your setup is correct. The PWM and encoder starts are inside the main loop, which means they're being repeatedly initialized every iteration. Maybe these should be moved outside the while loop, into the setup section? Does it maybe help if you restructure your main.cpp something like this:

```cpp
int main(void)
{
  /* USER CODE BEGIN 1 */
  /* USER CODE END 1 */

  /* MCU Configuration--------------------------------------------------------*/
  HAL_Init();
  SystemClock_Config();

  /* Initialize all configured peripherals */
  MX_GPIO_Init();
  MX_DMA_Init();
  MX_USART1_UART_Init();
  MX_TIM2_Init();
  MX_TIM3_Init();
  MX_TIM4_Init();
  MX_TIM5_Init();

  /* USER CODE BEGIN 2 */
  // Move timer initializations here, before setup()
  HAL_TIM_PWM_Start(&htim3, TIM_CHANNEL_1);
  HAL_TIM_PWM_Start(&htim4, TIM_CHANNEL_1);
  HAL_TIM_Encoder_Start(&htim2, TIM_CHANNEL_ALL); // Start encoder 1
  HAL_TIM_Encoder_Start(&htim5, TIM_CHANNEL_ALL); // Start encoder 2
  
  // Start timers with interrupts
  HAL_TIM_Base_Start_IT(&htim3);
  HAL_TIM_Base_Start_IT(&htim4);

  setup();
  /* USER CODE END 2 */

  /* Infinite loop */
  while (1)
  {
    loop();
  }
  /* USER CODE END 3 */
}
```

The "Tried to publish before configured topic id" error typically occurs when you're trying to publish messages before the ROS node has fully initialized and established communication with the rosserial server. This often happens in conjunction with sync issues.
Here's what I recommend trying, in order:

1. First, try the code rearrangement I suggested above - moving the timer initializations outside the main loop. This could help with the sync issues.
2. If you're still seeing issues, modify your setup() to ensure proper ROS initialization:

```cpp
void setup() {
  nh.initNode();
  // Add delay after node init
  HAL_Delay(1000);  // Give ROS time to initialize
  
  // Initialize all your publishers/subscribers
  nh.advertise(your_publisher);
  // Add another small delay
  HAL_Delay(100);
}

void loop() {
  // Your publishing code
  your_publisher.publish(&your_message);
  
  nh.spinOnce();
  HAL_Delay(10);  // Adjust this delay as needed
}
```

3. Common causes of both issues:
Node not fully initialized before publishing
Incorrect baudrate settings
Topics being published too quickly
Memory issues (if you have many publishers/subscribers)


Looking at your code in `mainpp.cpp`, potential causes for both the sync issues and the "publish before configured" error could be the following: 

- Timing Issues:

```cpp
void setup()
{
    base_controller.setup();
    base_controller.init();
    nh.loginfo("Setup finished");
    // Missing initialization delay and explicit node initialization
}
```

Try modifying the `setup()`:

```cpp
void setup()
{
    nh.initNode();  // Explicitly initialize the node
    HAL_Delay(1000);  // Give time for node initialization
    
    base_controller.setup();
    base_controller.init();
    
    HAL_Delay(100);  // Give time for topics to register
    nh.loginfo("Setup finished");
}
```

- Loop Timing:
Your loop contains several blocks of time-critical code but the delay might be too long:

```cpp
    nh.spinOnce();
    HAL_Delay(100);  // 100ms might be too long

Try reducing this to:

```cpp
    nh.spinOnce();
    HAL_Delay(10);  // Try a shorter delay
```

- Add error checking in the setup:

```cpp
void setup()
{
    nh.initNode();
    HAL_Delay(1000);
    
    if (!nh.connected()) {
        nh.logfatal("Failed to connect to ROS master");
        return;
    }
    
    base_controller.setup();
    base_controller.init();
    
    HAL_Delay(100);
    nh.loginfo("Setup finished");
}
```

- Consider checking connection status in the loop:

```cpp
void loop()
{
    if (!nh.connected()) {
        nh.spinOnce();
        HAL_Delay(1000);
        return;
    }
    
    // Rest of your existing loop code...
}
```

- Memory Usage:
You have several objects (controllers, encoders) that might be publishing topics. Try also not exceeding the maximum number of publishers/subscribers for your MCU's memory constraints.

- base controller code
The base controller setup looks generally ok, but there are a few potential improvements we could try:

```cpp
template <typename L298MotorController>
void diffbot::BaseController<L298MotorController>::setup()
{
    nh_.initNode();
    // Add delay after node init
    HAL_Delay(500);

    // Memory allocation checks
    if (!msg_measured_joint_states_.position || !msg_measured_joint_states_.velocity) {
        nh_.logfatal("Failed to allocate memory for joint states");
        return;
    }

    // Setup publishers first
    nh_.advertise(pub_encoders_);
    HAL_Delay(50);  // Small delay between publisher setups
    
    msg_measured_joint_states_.position = (double*)malloc(sizeof(double) * 2);
    msg_measured_joint_states_.position_length = 2;
    msg_measured_joint_states_.velocity = (double*)malloc(sizeof(double) * 2);
    msg_measured_joint_states_.velocity_length = 2;
    nh_.advertise(pub_measured_joint_states_);
    HAL_Delay(50);

    // Then setup subscribers
    nh_.subscribe(sub_wheel_cmd_velocities_);
    nh_.subscribe(sub_reset_encoders_);
    nh_.subscribe(sub_pid_left_);
    nh_.subscribe(sub_pid_right_);

    // Connection check with timeout
    uint32_t connection_timeout = HAL_GetTick() + 5000;  // 5 second timeout
    while (!nh_.connected())
    {
        nh_.spinOnce();
        HAL_Delay(10);
        
        if (HAL_GetTick() > connection_timeout) {
            nh_.logfatal("Failed to connect to ROS master within timeout");
            return;
        }
    }
    
    nh_.loginfo("Base controller setup completed successfully");
}
```

Key improvements:

- Added delays between initialization steps
- Added memory allocation checks
- Separated publisher and subscriber setup
- Added timeout to the connection check
- Added success log message

Two things to check:

- Are you freeing the allocated memory somewhere when needed?
- What is your UART baudrate set to? This needs to match on both the ROS and mcu side.

Finally, adding logging statements in the form `nh.loginfo("helpful message");` to appropriate places will help identify:

- If the setup process completes successfully
- Where in the initialization process things might be failing
- If and when the ROS connection is lost
- When control cycles are happening
- Encoder readings
- Don't log too frequently as it can affect performance

## Start with a simple example

If you haven't done it already I'd start with a minimal rosserial example on your STM32 mcu. Maybe getting a LED to switch on/off through a simple rostopic: http://wiki.ros.org/rosserial_arduino/Tutorials/Blink. Just to see if the basic setup works and then go back to the much harder to setup base_controller code.

Good luck and I hope some of these suggestions will help you 🤞

**Reply by thanhtruongphan** [reply]
because of my STM32, I don't know, I tested separately with rosserial_stm32 > publisher encoder ticks, It doesn't run smoothly, maybe all the error come from my stm32...
I just orderd a teensy board, I'll use it :'(
